src/RealtimeEffectPanel.cpp

```diff
@@ -1000,6 +1001,37 @@ RealtimeEffectPanel::RealtimeEffectPanel(
             break;
          }
    });
+
+   mUndoSubscription = UndoManager::Get(mProject).Subscribe(
+      [this](UndoRedoMessage message)
+      {
+         auto& trackList = TrackList::Get(mProject);
+
+         // Realtime effect UI is only updated on Undo or Redo
+         if (message.type == UndoRedoMessage::Type::UndoOrRedo)
+         {
+            for (auto waveTrack : trackList.Any<WaveTrack>())
+               UpdateRealtimeEffectUIData(*waveTrack);
+         }
+
+         // But mPotentiallyRemovedTracks processing happens as fast as possible.
+         // This event is fired right after the track is deleted, so we do not
+         // hold the strong reference to the track much longer than need.
+         if (mPotentiallyRemovedTracks.empty())
+            return;
+         
+         for (auto track : mPotentiallyRemovedTracks)
+         {
+            // By construction, track cannot be null
+            assert(track != nullptr);
+            auto matchingTrack = trackList.FindById(track->GetId());
+
+            if (matchingTrack == nullptr)
+               HideRealtimeUIForTrack(*track);
```

Comment:
This requires that #3365 be merged first, so that the dialog is closed when it must be.

Reply:
Yep, right now it is  kind of no-op

Reply:
Correct that, it is rather, still closing the dialog unnecessarily.

We want the dialog to close exactly when it should.

Reply:
Now I'm trying to make it work by cherry-picking my fix onto this branch, and it doesn't, revealing an incompleteness of my proposed fix.  So watch for an amendment of #3365 !

Reply:
Okay, I ameneded #3365, which you should review first.

But still not enough:  if I cherry-pick it onto your branch, I must make one more change:

Your message handler must do nothing messages of type `Purge`.

Now I observe that undo and redo close the dialog only exactly as needed.

So, are we done yet, debugging undo/redo?

No, I think there is one more problem.  Wait for more.